### PR TITLE
Add sphinx-builder `linkcheck`

### PR DESCRIPTION
**Pull request by LecrisUT**
Depends on: #2483

There are a bunch of linkcheck failures, many of which come from the auto-generated pages. Not sure how these should be handled

### CODE REVIEW

**Comment by LecrisUT** [comment]
Oh wow, this is what it expanded for me:
```rst
* Verified by `/tests/core/adjust <https://github.com/LecrisUT/tmt/tree/feat/linkcheck/tests/core/adjust/main.fmf>`_ 
```
It seems the generation automatically puts in the git repo and branch, but then on a PR, the git repo can be either target or source depending on the checkout. I think that is where it trips up :thinking:

**Comment by happz** [comment]
> @happz looks like we have a few more broken links being generated.

Yes, my https://github.com/teemtee/tmt/pull/2940 did not try to address everything. There's https://github.com/teemtee/tmt/issues/2868 for the links to tmt internals and tests as created by the template, which seems to be the vast majority of the remaining broken links.stories

> No idea about this one `( spec/core: line 131) broken None - ` Test docs, like `/tests/execute/results-custom.fmf` -> `tests/execute/result/custom`
> 
> Looking into how it's being generated.

Check out https://github.com/teemtee/tmt/issues/2868#issuecomment-2109611146. I'd be happy to look into fixing this, since it's the code I wrote and left in less than ideal state, but if you feel like diving in, be my guest :)

Today I'd go with the idea mentioned in the last paragraph, i.e. moving the macro away from the template, making it a full-fledged Python code & expose it back to the template, mostly because it can be tested easily. I'd add a couple of unit tests to verify it's rendering links correctly - should be mostly `Link` on the input, `str` on the output, therefore easy to parameterize with a pile of test cases. Feel free to ping me if you hit some dumb piece of code.

> Oh wow, this is what it expanded for me:
> 
> ```
> * Verified by `/tests/core/adjust <https://github.com/LecrisUT/tmt/tree/feat/linkcheck/tests/core/adjust/main.fmf>`_ 
> ```
> 
> It seems the generation automatically puts in the git repo and branch, but then on a PR, the git repo can be either target or source depending on the checkout. I think that is where it trips up 🤔

That's one issue for sure, there are more though, e.g. some links lead to `main.fmf` when the test is not defined in `main.fmf` and the file has a different name, and so on. Basically the macro tries to guess, and in some cases it's wrong.

**Comment by martinhoyer** [comment]
Thanks @happz!  

> if you feel like diving in, be my guest :)

I'll be afk for the rest of this week unfortunately. Also, I've familiarized myself with the templates a bit and think I get the general idea, but there is still plenty of things I'm seeing for the first time to catch-up with.

**Comment by happz** [comment]
With https://github.com/teemtee/tmt/pull/3001 applied, I get no broken links reported from the linkcheck builder.

**Comment by LecrisUT** [comment]
> With #3001 applied, I get no broken links reported from the linkcheck builder.

Let me rebase on your commit first just to make sure it works for forks as well

**Comment by happz** [comment]
> > With #3001 applied, I get no broken links reported from the linkcheck builder.
> 
> Let me rebase on your commit first just to make sure it works for forks as well

It's probably not going to, not out of the box, there's still an issue with forks and/or SSH config. I had to apply this to unblock my local environment:

```diff
diff --git a/tmt/utils.py b/tmt/utils.py
index 596c66b9..01493215 100644
--- a/tmt/utils.py
+++ b/tmt/utils.py
@@ -4291,6 +4291,8 @@ def web_git_url(url: str, ref: str, path: Optional[Path] = None) -> str:
     if path:
         url += str(path)

+    url = url.replace('github:teemtee', 'https://github.com/teemtee').replace('feat/linkcheck', 'main')
+
     return url

```

**Comment by happz** [comment]
Right, and the branch is also wrong, it's all `main` and `HEAD`, but that's another problem to resolve. It works now, it won't work when PR starts pointing at new things.

**Comment by LecrisUT** [comment]
Hmm, but looking at the original generated links it actually looked fine:
```rst
* Verified by `/tests/core/adjust <https://github.com/LecrisUT/tmt/tree/feat/linkcheck/tests/core/adjust/main.fmf>`_ 
```

It points to a correct file, so should it actually have that replace?

**Comment by happz** [comment]
> Hmm, but looking at the original generated links it actually looked fine:
> 
> ```
> * Verified by `/tests/core/adjust <https://github.com/LecrisUT/tmt/tree/feat/linkcheck/tests/core/adjust/main.fmf>`_ 
> ```
> 
> It points to a correct file, so should it actually have that replace?

Well, even better then :) Things did not work for me locally, hence the `replace()`, if you're fine with how they work in Github check, fewer things to worry about for me :)

it might have been the SSH config rather than the fork that was ruining my local work.

**Comment by happz** [comment]
Anyway, I don't know where the single broken link comes from - I can't reproduce it:

```
2024-06-10T08:13:19.5300673Z (       spec/core: line  131) broken    None - 
```

In my generated sources, it's this line:

```
* Verified by `/tests/core/adjust <github:teemtee/tmt/tree/docs-improve-rendering-of-story-links/tests/core/adjust/main.fmf>`__ 
```

**Comment by martinhoyer** [comment]
> Anyway, I don't know where the single broken link comes from - I can't reproduce it:

It's driving me crazy.  
btw, shouldn't we see test links ending with `.fmf` in the output as `ok`?

**Comment by LecrisUT** [comment]
Sorry, should have done this long ago, let's what exactly is being produced in github action: https://github.com/teemtee/tmt/actions/runs/9450580783/artifacts/1585890004

**Comment by martinhoyer** [comment]
> Sorry, should have done this long ago, let's what exactly is being produced in github action: https://github.com/teemtee/tmt/actions/runs/9450580783/artifacts/1585890004
> It is literally `None` 😅. Seems to be `{{ url }}` expansion

Thanks! Indeed. I'm confused why the doc build is successful. Perhaps we should double-check the build command arguments?

**Comment by LecrisUT** [comment]
Oh, yeah, I didn't even realize the normal build actually worked perfectly fine. One difference is that `hatch` is used for the Github actions.

**Comment by LecrisUT** [comment]
Ok, I probably have a clue:
```rst
..
    [Debug] emit_tmt_object_links(link).object = /tests/core/adjust [Test]
    fmf_id.url = https://github.com/teemtee/tmt
    fmf_id.ref = None
```
It's probably that `actions/checkout` on PRs is pointing to the merge commit [^1] and it seems `fmf_id.ref` is not able to account for that. RTD uses source commit instead.

[^1]: https://github.com/actions/checkout?tab=readme-ov-file#checkout-pull-request-head-commit-instead-of-merge-commit

**Comment by happz** [comment]
Interesting. It seems to me that `ref` is also `HEAD` in RTD PR job, no idea why. At least `HEAD` is propagated into links in docs generated for a PR.

**Comment by LecrisUT** [comment]
Maybe, it's more simple than that, just the command failing at:
```python
    branch = run(Command("git", "rev-parse", "--abbrev-ref", "--symbolic-full-name", "@{u}"))
```
Maybe there are better commands than `rev-parse` to get what is basically:
```console
$ git checkout --progress --force refs/remotes/pull/2489/merge
```

The difference with RTD is that it uses:
```console
git fetch origin --force --prune --prune-tags --depth 50 pull/2489/head:external-2489
```
(Note the : making it a named branch, vs the GH action)

**Comment by LecrisUT** [comment]
@happz checkout 94fc226ca5d757ff63052119477fc57712068152 and b9d84acab0ced825e8e4b62ce131b4add431f6f0 if you've got time. This almost works, just need to sanitize `web_git_url` a bit more to be able to read and navigate `ref=refs/remotes/pull/N/{merge,head}` (and equivalent for gitlab) and translate it to PR-like url. However, I don't seem to find a `tree` url to navigate a PR :/

File: .github/workflows/doc-tests.yml
Comment (thrix) [inline .github/workflows/doc-tests.yml:27]: @LecrisUT  what is this good for exactly?

File: .github/workflows/doc-tests.yml
Comment (thrix) [inline .github/workflows/doc-tests.yml:46]: Can you tell me why we are moving from hatch to pip?

**Comment by thrix** [comment]
@LecrisUT we talked about this PR today, could you maybe rebase it and write down the current state, or make it a draft if still in works pls?

File: .github/workflows/doc-tests.yml
Comment (LecrisUT) [reply .github/workflows/doc-tests.yml:27]: Sorry, these are from a bunch of debug commits that I've added to debug: https://github.com/teemtee/tmt/pull/2489#issuecomment-2158802072 i.e. how to get the links to point to PR files. Will remove those, but need some feedback on:
```console
$ git for-each-ref --format="%(refname)" --points-at HEAD
```

See the github action log on what everything is expanded to. For PRs `refs/remotes/pull/2489/merge` works well (replace that in the commitish) and anything that can expand to that is a good approach (needs to expand to a `ref` that is recognized on the target repo)

File: .github/workflows/doc-tests.yml
Comment (LecrisUT) [reply .github/workflows/doc-tests.yml:46]: I think this is a leftover from the initial attempt, will change it back to use `hatch`

**Comment by martinhoyer** [comment]
Any ideas why it failed @LecrisUT @happz ?
It seems to be this line:
```
(       spec/core: line  131) broken    None - 
```
, which if I understand it correctly should be https://github.com/teemtee/tmt/tree/HEAD/tests/core/adjust/main.fmf link, which works fine on rtd latest.

**Comment by LecrisUT** [comment]
The failure is because it is run on a fork, these are resolved in #3108

**Comment by martinhoyer** [comment]
> The failure is because it is run on a fork, these are resolved in #3108

If it's because of fork, wouldn't it break more links than just one?

**Comment by LecrisUT** [comment]
I don't quite remember, but there were a bunch of broken links inside the rst generated file, but it so happened that linkcheck aggregated the report or something, don't quite remember.

On RTD it showed fine because it uses a different git checkout approach and it doesn't fail here:
https://github.com/teemtee/tmt/blob/5fbbd89f8be22fdbb7efd7c7044ea0dd4d077ca7/tmt/utils/__init__.py#L4318-L4325

But git checkout in GH actions checksout the merge commitand that's where it breaks. Check the log of the checkout action to replicate locally.

**Review by thrix — APPROVED** [review]
LGTM

**Review by martinhoyer — APPROVED** [review]

**Review by therazix — APPROVED** [review]

**Review by seberm — APPROVED** [review]

**Comment by happz** [comment]
Hmm, blocked by https://github.com/teemtee/tmt/pull/3108 :/

**Comment by LecrisUT** [comment]
> Hmm, blocked by #3108 :/

Or maybe superseded-by since that one contains the commit from here

**Comment by psss** [comment]
> Or maybe superseded-by since that one contains the commit from here

Let's keep the `linkcheck` change here and remove it from #3108.

**Review by psss — APPROVED** [review]

**Comment by psss** [comment]
The extended unit tests failure will be fixed by #3227. Remaining failure is irrelevant.

**Pull request merged**
